### PR TITLE
Move config type creation from docker.utils functions to classes in docker.types

### PR DESCRIPTION
**Pull request by shin-**
Mostly internal API changes, but some of those are breaking (public) changes 
- `create_host_config` for some users who have not migrated to using the method through `APIClient`
- `create_ipam_config` and `create_ipam_pool`

Those functions now raise a `docker.errors.DeprecatedMethod` error pointing to the new classes.

This also fixes some egregious circular imports between `docker.types` and `docker.utils`

### CODE REVIEW

**Comment by bfirsh** [comment]
:+1:

Test failures, though.

File: docker/api/network.py [hunk 1]
```diff
@@ -47,7 +47,7 @@ def create_network(self, name, driver=None, options=None, ipam=None,
             driver (str): Name of the driver used to create the network
             options (dict): Driver options as a key-value dictionary
             ipam (dict): Optional custom IP scheme for the network.
-                Created with :py:meth:`~docker.utils.create_ipam_config`.
+                Created with :py:class:`~docker.types.IPAMConfig`.
```
Comment (bfirsh) [inline docker/api/network.py]: This won't actually link to anywhere unless they are represented in a docs file somewhere.

File: docs/api.rst [hunk 2]
```diff
@@ -49,15 +49,6 @@ Networks
   :members:
   :undoc-members:
 
-Utilities
```
Comment (bfirsh) [inline docs/api.rst]: Perhaps we could replace this with a "types" section to actually document them? (See my other comments.) Ditto the `HostConfig` object, and so on.

**Comment by bfirsh** [comment]
Looks good apart from docs!

**Comment by shin-** [comment]
Thanks Ben! Updated a lot of doc-related stuff and rebased, should look good now.

File: docker/types/services.py [hunk 3]
```diff
@@ -10,19 +11,15 @@ class TaskTemplate(dict):
...
-      the service. See the :py:class:`~docker.types.services.DriverConfig`
-      class for details.
-    * resources (dict): Resource requirements which apply to each individual
-      container created as part of the service. See the
-      :py:class:`~docker.types.services.Resources` class for details.
-    * restart_policy (dict): Specification for the restart policy which applies
-      to containers created as part of this service. See the
-      :py:class:`~docker.types.services.RestartPolicy` class for details.
-    * placement (list): A list of constraints.
+        container_spec (ContainerSpec): Container settings for containers
```
Comment (bfirsh) [inline docker/types/services.py]: A tip, but not important: these types can be links, like `` :py:class:`~docker.types.services.ContainerSpec` ``.

**Comment by bfirsh** [comment]
LGTM! :tada:

**Pull request merged**